src/i_winmusic.c

```diff
@@ -199,49 +200,43 @@ static void CALLBACK MidiStreamProc(HMIDIOUT hMidi, UINT uMsg,
     }
 }
 
-// Allocates the buffer and prepares the MIDI header. Set during initialization
-// by the main thread. BUFFER_INITIAL_SIZE should be large enough to avoid
-// reallocation by the MIDI thread during playback, due to a known memory bug
-// with midiOutUnprepareHeader() (detected by ASan). The calling thread must
-// have exclusive access to the shared resources in this function.
+// Reset buffer position and unprepare MIDI header. The calling thread must have
+// exclusive access to the shared resources in this function.
 
-static void PrepareHeader(void)
+static void ResetBuffer(void)
 {
     MIDIHDR *hdr = &MidiStreamHdr;
     MMRESULT mmr;
 
-    hdr->lpData = (LPSTR)buffer.data;
-    hdr->dwBytesRecorded = 0;
-    hdr->dwBufferLength = buffer.size;
-    mmr = midiOutPrepareHeader((HMIDIOUT)hMidiStream, hdr, sizeof(MIDIHDR));
+    if (!buffer.prepared)
+    {
+        return;
+    }
+
+    // Avoid ASan detection. Commentary by Microsoft: "It looks like
+    // midiOutPrepareHeader() allocates with HeapAlloc(), and then
+    // midiOutUnprepareHeader() deallocates with GlobalFree(GlobalHandle
+    // (...)). By design, this kind of allocator mismatch is an issue that ASan
+    // is designed to catch. It is theoretically possible for us to support
+    // this kind of code, but it’s not very high priority since it is undefined
+    // behavior, though it happens to work right now outside of ASan."
+
+#ifndef __SANITIZE_ADDRESS__
+    mmr = midiOutUnprepareHeader((HMIDIOUT)hMidiStream, hdr, sizeof(MIDIHDR));
     if (mmr != MMSYSERR_NOERROR)
     {
-        MidiError("midiOutPrepareHeader", mmr);
+        MidiError("midiOutUnprepareHeader", mmr);
     }
+#endif
+
```

Comment:
To be safe, add something like `memset(hdr, 0, sizeof(*hdr));` here or, at the very least, in `I_WIN_ShutdownMusic()`. This ensures the midi header is completely reset.

Reply:
I decided to do the `memset` in `StreamOut()` instead, that's what ZMusic does. After looking at the `MIDIHDR` structure, I'm not sure if we should nullify it:
```c
/* MIDI data block header */
typedef struct midihdr_tag {
    LPSTR       lpData;               /* pointer to locked data block */
    DWORD       dwBufferLength;       /* length of data in data block */
    DWORD       dwBytesRecorded;      /* used for input only */
    DWORD_PTR   dwUser;               /* for client's use */
    DWORD       dwFlags;              /* assorted flags (see defines) */
    struct midihdr_tag far *lpNext;   /* reserved for driver */
    DWORD_PTR   reserved;             /* reserved for driver */
#if (WINVER >= 0x0400)
    DWORD       dwOffset;             /* Callback offset into buffer */
    DWORD_PTR   dwReserved[8];        /* Reserved for MMSYSTEM */
#endif
} MIDIHDR, *PMIDIHDR, NEAR *NPMIDIHDR, FAR *LPMIDIHDR;
```
It's from `mmeapi.h` in Windows SDK.

Reply:
Alternatively, could `MidiStreamHdr` be changed to a pointer? Seems weird to carry over a potentially dirty `MIDIHDR` from one device to the next.

Rough example:
```c
// Replaces MidiStreamHdr:
static MIDIHDR *hdr;

// I_WIN_InitMusic():
hdr = calloc(1, sizeof(*hdr));

// I_WIN_ShutdownMusic():
free(hdr);
hdr = NULL;
```

Reply:
We don't carry it, we nullify it before using it in the `StreamOut()` function.
ZMusic has 100 times more users, so I think it's reasonable to copy their approach. However, GZDoom has long used Fluidsynth by default, so it's hard to tell how many users use midiStream() in ZMusic.

Reply:
You're right, that approach is fine.